dnadoh/cache.py

```diff
@@ -0,0 +1,118 @@
+
+import filecmp
+import requests
+import shutil
+import validators
+
+from pathlib import Path
+from urllib.parse import urljoin
+
+
+CONFIG = {
+    "cache": Path(__file__).parent.parent.joinpath("cache"),  # prints DNA-doh rootdir + cache
+    "remote": Path(__file__).parent.parent.joinpath("remote"),
+}
+
+
+class FileCache:
+    """Cache large remote files."""
+
+    def __init__(self, cache_dir=None, remote_url=None):
+        if cache_dir:
+            self.cache_dir = cache_dir
+        else:
+            self.cache_dir = CONFIG["cache"]
+        
+        if remote_url:
+            self.remote_url = remote_url
+        else:
+            self.remote_url = CONFIG["remote"]
+        
+        self.files = {}
+        
+    def get(self, filename):
+        """Return path to cached copy of file, getting as needed."""
+        if filename not in self.files:
+            # check if valid URL, which implies it can be downloaded
+            if validators.url(filename) is True:
+                # extact filename from URL
+                extracted_filename = filename.split('/')[-1]
+                cache_path = Path(self.cache_dir, extracted_filename)
+                if self._download_file(filename, cache_path):
+                    self.files[filename] = cache_path
+                else:
+                    # download failed
+                    print(f"{filename} could not be downloaded.")
+            else:
+                cache_path = Path(self.cache_dir, filename)
+                self._get_file(filename, cache_path)
+                self.files[filename] = cache_path
+        else:
+            # 4 - do a further check that the file is actually the same
+            if filecmp.cmp(Path(self.cache_dir, filename), Path(self.remote_url, filename)) is True:  # this doesn't work for remote URLs
+                return self.files[filename]
+            else:
+                self._get_file(filename, cache_path)
+                self.files[filename] = cache_path
+
+    def clear(self):
+        """Clear the cache, deleting local files."""
+        self.files.clear()
+        for filename in self.cache_dir.iterdir():
+            filename.unlink()
+
+    def _get_file(self, remote_path, local_path):
+        """Simulate getting a remote file."""
+        remote_path = Path(self.remote_url, remote_path)
+        # print(remote_path)
+        assert remote_path.exists(), f"{remote_path} does not exist"
+        shutil.copyfile(remote_path, local_path)
+
+    def _download_file(self, remote_url, local_path):
+        """Downloads a file from a remote url"""
+        # restrict download size - set 250 MB limit
+        h = requests.head(remote_url, allow_redirects=True)
+        header = h.headers
+        content_length = header.get("content-length", None)
+        # print(header)
+        # i think content length only works for txt files? couldn't see one with an image
+        if content_length and content_length > 2.5e-8:  # this is in bytes
```

Comment:
you're checking to see if the file is larger than 25 nanobytes (minus sign)

Reply:
even with the `-` taken out, the content length limit should be promoted to a constant at the top of the file: no magic numbers...

Reply:
 🤦🏻‍♀️ to accidentally adding that `-`

Reply:
"Physicists worry about decimal places, astronomers worry about exponents, and economists are happy if they get the sign right."